### PR TITLE
os/bluestore: fix bluefs log growth.

### PR DESCRIPTION
**Pull request by aclamk**
This is fix for issue 
https://tracker.ceph.com/issues/45903
https://bugzilla.redhat.com/show_bug.cgi?id=1821133

Original problem stemmed from BlueFS inability to replay log, which was caused by BlueFS previously wrote replay log that was corrupted, which was caused by BlueFS log growing to extreme size (~600GB), which was caused by OSD working in a way, when BlueFS::sync_metadata was never invoked.
This was possible only because OSD was working under extremely low write load, and single RocksDB WAL was purged for more then 14 days.

### CODE REVIEW

**Comment by ifed01** [comment]
@aclamk - did you see my comment in https://tracker.ceph.com/issues/45903?
This ticket is for Luminous release while the former lacks a patch making bluefs log compaction more aggressive, see https://github.com/ceph/ceph/pull/34876

**Comment by ifed01** [comment]
Additionally for the last couple of weeks I've seen another(?) bluefs log issue which looks related - under some circumstances (heavily fragmented and/or massive space allocation) BlueFS might overflow its log runaway (4MB) and either assert with "ceph_assert(h->file->fnode.ino != 1)" or even corrupt the log. See https://tracker.ceph.com/issues/45519
I'm still not 100% aware whether all this stuff is related.

**Comment by ifed01** [comment]
@aclamk - wondering if you were aware of #34876 and did all the analysis for master not luminous? In other words - are you changes needed when #17354/#34876 are in place?

**Comment by aclamk** [comment]
@ifed01 
I have has a case when BlueFS::sync_metadata was never invoked.
I think that https://github.com/ceph/ceph/pull/34876 does not cover this situation.

**Comment by aclamk** [comment]
@ifed01 
I agree that https://tracker.ceph.com/issues/45519 is related to https://tracker.ceph.com/issues/45903.
In my observations actual corruption of BlueFS log can happen earlier then detection of runway overflow; but truly, having highly fragmented BlueFS files can make log grow very quickly.

File: src/os/bluestore/BlueFS.h [hunk 1]
```diff
@@ -558,16 +560,19 @@ class BlueFS {
   void handle_discard(unsigned dev, interval_set<uint64_t>& to_release);
 
   void flush(FileWriter *h) {
-    std::lock_guard l(lock);
+    std::unique_lock l(lock);
     _flush(h, false);
+    _maybe_compact_log(l);
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.h:565]: This looks like an overkill to me -IIUC flush doesn't guarantee data persistence and hence KV has to use FSync from time to time. I.e. compacting log from fsync is likely to be enough. Besides I'm curious about performance impact of this call on each flush.

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -2622,6 +2624,10 @@ int BlueFS::_flush_and_sync_log(std::unique_lock<ceph::mutex>& l,
 
   logger->inc(l_bluefs_logged_bytes, bl.length());
 
+  if (just_expanded_log) {
+    ceph_assert(bl.length() <= runway); // if we write this, we will have an unrecoverable data loss
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueFS.cc:2628]: No, it is deliberately checking previous size of runway. New size of runway will likely be enough to accommodate writes to log, but if we have not enough space to fit in *previous* extents, then replaying extension will not be possible.

[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueFS.cc:2628]: No, it is deliberately checking previous size of runway. New size of runway will likely be enough to accommodate writes to log, but if we have not enough space to fit in previous extents, then replaying extension will not be possible.

File: src/os/bluestore/BlueFS.h (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueFS.h]: @ifed01 Modified to check only when flush only really flushed something to disk

Comment (ifed01) [reply src/os/bluestore/BlueFS.h]: @aclamk - but I'm still curious if compacting log on fsync only would be enough? Does RocksDB call fsync often enough during that infrequent writes processing which you're referring in the description.

Comment (aclamk) [reply src/os/bluestore/BlueFS.h]: @ifed01 Customer had a cluster where there was almost no write traffic. Only op was write to WAL for 23 bytes once every second. Unfortunately, it translated to 109 kB of BlueFS log per second (mostly due to BlueFS log containing >70000 extents).

**Comment by aclamk** [comment]
@ifed01 I rate that this PR as partial fix to #45903, because it is still possible for bluefs log to grow so much, that it will be stopped by assert preventing log corruption. Although is of course very unlikely to happen, but final solution should compact log immediately if such condition is detected.

**Comment by aclamk** [comment]
@ifed01 I changed conf setting in test_bluefs.cc, I asked for review because of that.

File: src/test/objectstore/test_bluefs.cc [hunk 3]
```diff
@@ -54,6 +55,36 @@ class TempBdev {
...
+      saved_settings.pop();
+    }
+    conf.set_safe_to_start_threads();
+    conf.apply_changes(nullptr);
+  }
+  void SetVal(const char* key, const char* val) {
+    std::string skey(key);
+    std::string prev_val;
+    conf.get_val(skey, &prev_val);
+    conf.set_val_or_die(key, val);
```
Comment (ifed01) [inline src/test/objectstore/test_bluefs.cc:79]: nit: use skey not key here

File: src/os/bluestore/BlueFS.cc [hunk 4]
```diff
@@ -2940,11 +2940,24 @@ void BlueFS::wait_for_aio(FileWriter *h)
 }
 #endif
 
-int BlueFS::_flush(FileWriter *h, bool force)
+int BlueFS::_flush(FileWriter *h, bool force, std::unique_lock<ceph::mutex>& l)
+{
+  bool flushed;
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueFS.cc:2945]: @ifed01 thanks, I want to fix all nits!

File: src/test/objectstore/test_bluefs.cc (hunk 3 above)
Comment (aclamk) [reply src/test/objectstore/test_bluefs.cc:79]: thx

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-06-23_07:49:50-rados-wip-kefu-testing-2020-06-23-1007-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-06-23_15:15:59-rados-wip-kefu-testing-2020-06-23-1353-distro-basic-smithi/

---

- http://pulpito.ceph.com/kchai-2020-06-23_07:49:50-rados-wip-kefu-testing-2020-06-23-1007-distro-basic-smithi/5171960, no clues, osd.2 failed to reply queries from osd.0 when peering. but there was no clues in its log.
- http://pulpito.ceph.com/kchai-2020-06-23_15:15:59-rados-wip-kefu-testing-2020-06-23-1353-distro-basic-smithi/5173648:  tracked by https://tracker.ceph.com/issues/45721
- http://pulpito.ceph.com/kchai-2020-06-23_15:15:59-rados-wip-kefu-testing-2020-06-23-1353-distro-basic-smithi/5173654/: no tests failed..

**Comment by tchaikov** [comment]
jenkins test docs

**Pull request merged**